contracts/epochs/FlowEpoch.cdc

```diff
@@ -432,6 +497,234 @@ access(all) contract FlowEpoch {
+                message: "recovery epoch counter should equal current epoch counter + 1"
+            )
+            self.stopEpochComponents()
+            let randomSource = FlowEpoch.generateRandomSource()
+
+            let numViewsInStakingAuction = FlowEpoch.configurableMetadata.numViewsInStakingAuction
+            let numViewsInDKGPhase = FlowEpoch.configurableMetadata.numViewsInDKGPhase
+
+            /// Create new EpochMetadata for the recovery epoch with the new values
+            let newEpochMetadata = EpochMetadata(
+            /// Increment the epoch counter when recovering with a new epoch
+            counter: FlowEpoch.proposedEpochCounter(),
+            seed: randomSource,
+            startView: startView,
+            endView: endView,
+            stakingEndView: stakingEndView,
+            // The following fields will be overwritten in `calculateAndSetRewards` below
+            totalRewards: 0.0,
+            collectorClusters: [],
+            clusterQCs: [],
+            dkgKeys: dkgPubKeys)
+
+            /// Save the new epoch meta data, it will be referenced as             
+            /// the epoch progresses through each phase.
+            FlowEpoch.saveEpochMetadata(newEpochMetadata)
+
+            /// Calculate rewards for the current epoch
+            /// and set the payout for the next epoch
+            FlowEpoch.calculateAndSetRewards()
+
+            /// emit the epoch recover event
+            self.emitEpochRecoverEvent(epochCounter: FlowEpoch.proposedEpochCounter(),
+                startView: startView,
+                stakingEndView: stakingEndView,
+                endView: endView,
+                numViewsInStakingAuction: numViewsInStakingAuction,
+                numViewsInDKGPhase: numViewsInDKGPhase,
+                nodeIDs: nodeIDs,
+                clusterAssignments: clusterAssignments,
+                randomSource: randomSource,
+                targetDuration: targetDuration,
+                targetEndTime: targetEndTime,
+                clusterQCVoteData: clusterQCVoteData,
+                dkgPubKeys: dkgPubKeys,
+            )
+
+            /// Start a new Epoch, which increments the current epoch counter
+            FlowEpoch.startNewEpoch()
+        }
+
+        /// Ends the currently active epoch and restarts it with the provided configuration.
+        /// This function is intended to update the current epoch configuration when a recovery
+        /// transaction needs to be submitted more than once. This function differs 
+        /// from recoverNewEpoch because it does not increment the epoch counter. It also 
+        /// does not calculate and set rewards avoiding double paying rewards for the same epoch.
+        /// This meta data will be emitted in the EpochRecover service event. This function is used 
+        /// within sporks to recover the network from Epoch Fallback Mode (EFM).
+        access(all) fun recoverCurrentEpoch(startView: UInt64,
```

Comment:
Are we sure we shouldn't combine this and `recoverNewEpoch` into a single function? Might save some lines of code and we could potentially just detect which one we need. I don't know for sure though

Reply:
Yes that how it initially was but we decided to make them 2 distinct functions so that it is more obvious the difference in behavior.